lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -499,30 +442,19 @@ handleGetBlocks logic oq = listenerConv oq $ \__ourVerInfo nodeId conv -> do
             mgb nodeId
         mHashes <- getBlockHeaders' logic mgbFrom mgbTo
         case mHashes of
-            Right (Just hashes) -> do
+            Right hashes -> do
                 logDebug $ sformat
                     ("handleGetBlocks: started sending "%int%
                      " blocks to "%build%" one-by-one: "%listJson)
                     (length hashes) nodeId hashes
                 for_ hashes $ \hHash ->
                     getBlock logic hHash >>= \case
-                        Right (Just b) -> send conv (MsgBlock b)
-                        -- TODO handle Left *and* Right Nothing
-                        _ -> do
-                            send conv (MsgNoBlock $
-                                       "Couldn't retrieve block with hash " <> pretty hHash)
-                            --failMalformed
+                        Just b -> send conv $
+                            MsgBlock b
+                        Nothing  -> send conv $
+                            MsgNoBlock ("Couldn't retrieve block with hash " <> pretty hHash)
                 logDebug "handleGetBlocks: blocks sending done"
             _ -> logWarning $ "getBlocksByHeaders@retrieveHeaders returned Nothing"
-  {-
-  where
-    FIXME this makes no sense. Diffusion layer should not make judgements about
```

Comment:
Arguable. The fact db says `fromTo a b = [a..b]` and then `exists c in [a..b] which is not in db` is enough to `throwM DBMalformed`. This logic should be restored.

Reply:
This is an unfortunate artefact of a poor design: `DB.getBlockHeaders'` and `DB.getBlock` are completely divorced; there's no apparent connection between them, and so this opportunity for an error arises. Worse still: it's the *caller*'s responsibility to decide what the error is, even though the *callee* (database program) is in a much better position to do so!

Here's an idea for a better design: instead of pulling in a list of headers, then getting the block for each, stream the blocks directly from the headers.

```Haskell
blocks :: HeaderHash -> HeaderHash -> Producer m Block
```

Then the `Producer` (conduit or pipe or whatever) can throw the malformed exception, and the *caller* here doesn't have to take on that burden. What do you think?

Reply:
I think that in this particular place using streaming is a much better decision. But, in general, some other code may want to use `getBlockHeaders` without using `getBlocks` right after. I assume it's safe to rewrite it using conduits for now.

Reply:
I'll put back the db malformed errors. We'll probably implement a streaming style with CSL-2148